### ISSUE TITLE
How do I react to ONLY silent token refreshes?

### ISSUE THREAD

**Issue by mikebper** [AzureAD/microsoft-authentication-library-for-js#5902]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.31.0

### Wrapper Library

MSAL Angular (@azure/msal-angular)

### Wrapper Library Version

2.4.6

### Public or Confidential Client?

Public

### Description

How do I do something when tokens are refreshed, and only when tokens are refreshed?

I need to be able to do this in order to get the AccountInfo data from the new token. I do not want this code executing at any other time (E.G., at login, or when retrieving a token from cache), because I already have the AccountInfo data from the login response.

I read the [event](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/events.md) docs and tried numerous approaches (outlined below), and I have been unable to identify a way to do this.

### MSAL Configuration

```javascript
app.module.ts:
MsalModule.forRoot(new PublicClientApplication(msalConfig),
    {
      interactionType: InteractionType.Redirect,
      authRequest: {
        scopes: protectedResource.authRequest.scopes,
        redirectUri: '/'
      }
    },
    {
      interactionType: InteractionType.Redirect,
      protectedResourceMap: new Map([
        [protectedResource.myApiEndpoint.endpoint, protectedResource.myApiEndpoint.scopes],
      ])
    })
],
providers: [    
  {
    provide: HTTP_INTERCEPTORS,
    useClass: MsalInterceptor,
    multi: true
  },
  MsalGuard
],
bootstrap: [
  AppComponent,
  MsalRedirectComponent
]

auth-config.ts:
import { LogLevel, Configuration, BrowserCacheLocation } from '@azure/msal-browser';
import { environment } from 'src/environments/environment';

export const b2cPolicies = {
    authorities: {
        signIn: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_SIGNIN',
        },
        resetPassword: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_PASSWORDRESET',
        },
        editProfile: {
            authority: 'https://myapib2cdev.b2clogin.com/myapib2cdev.onmicrosoft.com/B2C_1A_PROFILEEDIT'
        }
    },
    authorityDomain: 'myapib2cdev.b2clogin.com'
};

export const msalConfig: Configuration = {
    auth: {
        clientId: 'my-client-ID',
        authority: b2cPolicies.authorities.signIn.authority,
        knownAuthorities: [b2cPolicies.authorityDomain],
        redirectUri: '/',
        postLogoutRedirectUri: '/'
    },
    cache: {
        cacheLocation: BrowserCacheLocation.SessionStorage
    },
    system: {
        loggerOptions: {
            logLevel: LogLevel.Error,
            loggerCallback: (level, message, containsPii) => {
                if (containsPii) {
                    return;
                }
                switch (level) {
                    case LogLevel.Error:
                        console.error(message);
                        return;
                    case LogLevel.Info:
                        console.info(message);
                        return;
                    case LogLevel.Verbose:
                        console.debug(message);
                        return;
                    case LogLevel.Warning:
                        console.warn(message);
                        return;
                }
            },
            piiLoggingEnabled: false
        }
    }
}

export const protectedResource = {
    authRequest: {
        scopes: ['https://myapib2cdev.onmicrosoft.com/myapi/Read']
    },
    myApiEndpoint: {
        endpoint: `${environment.myApiBaseUrl}/myApiEndpoint`,
        scopes: ['https://myapib2cdev.onmicrosoft.com/myapi/Read']
    },
}
```

### Relevant Code Snippets

I tried subscribing to the below with the indicated results:

1) Does not get called when acquireTokenSilent occurs:

```typescript
    this.msalBroadcastService.msalSubject$
      .pipe(
        filter((msg: EventMessage) => msg.interactionType == InteractionType.Silent && msg.eventType === EventType.ACQUIRE_TOKEN_BY_CODE_SUCCESS),
        takeUntil(this.destroying)
      )
```

2) Gets called after login when calling protected resource for the first time, even though no request is made to the token endpoint at that time (token was already acquired at login--no refresh has occurred yet):

```typescript
    this.msalBroadcastService.msalSubject$
      .pipe(
        filter((msg: EventMessage) => msg.interactionType == InteractionType.Silent),
        takeUntil(this.destroying)
      )
```

3) Gets called after login when calling protected resource for the first time, even though no request is made to the token endpoint at that time (token was already acquired at login--no refresh has occurred yet):

```typescript
    this.msalBroadcastService.msalSubject$
      .pipe(
        filter((msg: EventMessage) => msg.eventType === EventType.ACQUIRE_TOKEN_SUCCESS),
        takeUntil(this.destroying)
      )
```

4) Does not get called when token refresh occurs:

```typescript
    this.msalBroadcastService.inProgress$
      .pipe(
        filter((status: InteractionStatus) => status === InteractionStatus.None),
        takeUntil(this.destroying)
      )
```

### Identity Provider

Azure B2C Custom Policy

### Source

External (Customer)

**Comment by mikebper** [comment]
I think that either:

1. For `filter((status: InteractionStatus) => status === InteractionStatus.None)`, an 'InteractionStatus.None' message should be broadcast after tokens have refreshed via acquireTokenSilent, so that all logic can be centralized there.

or

2. For `filter((msg: EventMessage) => msg.interactionType == InteractionType.Silent && msg.eventType === EventType.ACQUIRE_TOKEN_BY_CODE_SUCCESS)`, this should not be triggered when the first protected resource is requested after having logged in despite no silent request actually being made to the token endpoint.

Preferably option 1.

**Comment by tnorling** [comment]
Something like this should work:

```js
this.msalBroadcastService.msalSubject$
      .pipe(
        filter((msg: EventMessage) => {
          if (msg.eventType === EventType.ACQUIRE_TOKEN_SUCCESS) {
            const payload = msg.payload as AuthenticationResult;
            return !payload.fromCache;
          } else {
            return false;
          }
        }),
        takeUntil(this.destroying)
      )
```

Note that `acquireTokenByCode` is a different API from `acquireTokenSilent` which is why `ACQUIRE_TOKEN_BY_CODE_SUCCESS` isn't being fired for you.

**Comment by mikebper** [comment]
I confirm that this works. Thank you! I think I did ACQUIRE_TOKEN_BY_CODE_SUCCESS as one of my tests from an autocomplete accident, but thanks for the explanation. :)  Maybe a new EventType could be added at some point.  Something like:

ACQUIRE_TOKEN_FROM_IDP_SUCCESS


The other issue I opened is still occurring (token refreshing every request after first time token endpoint is called, with AAD social login users only).
